### PR TITLE
feature(sct): add support for using dns names in cluster

### PR DESCRIPTION
**Pull request by soyacz**
We want to test if Scylla works with dns names instead of IP addresses.

Adding proper methods and conditions in various places according to test configuration.

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by soyacz** [comment]
basic is working: https://argus.scylladb.com/workspace?state=WyJjMzhkMjA1MC0xZTEyLTQ3OWEtODk0Mi00MGNmYWFlNDQ4NTkiXQ
Next job I run with lower load and longer duration to try as many nemesis as possible.

File: sdcm/cluster.py [hunk 1]
```diff
@@ -3939,7 +3950,10 @@ def set_seeds(self, wait_for_timeout=300, first_only=False):
 
     @property
     def seed_nodes_ips(self):
```
Comment (fruch) [inline sdcm/cluster.py:3952]: maybe we need to rename this function ? since it doesn't return ips only anymore ?

File: sdcm/cluster.py [hunk 2]
```diff
@@ -792,6 +793,14 @@ def public_ip_address(self) -> Optional[str]:
             self._public_ip_address_cached = self._get_public_ip_address()
         return self._public_ip_address_cached
 
+    @property
+    def public_dns_name(self) -> str:
+        return self.name
```
Comment (fruch) [inline sdcm/cluster.py:798]: i'm not sure we'll want to default it like that...

not if we didn't tested it for GCP/Azure/docker

maybe raising not implemented is better for now ?

File: sdcm/sct_config.py [hunk 3]
```diff
@@ -1473,6 +1473,9 @@ class SCTConfiguration(dict):
              help="""Forces GossipingPropertyFileSnitch (regardless `endpoint_snitch`) to simulate racks.
              Provide number of racks to simulate."""),
 
+        dict(name="use_dns_names", env="SCT_USE_DNS_NAMES", type=boolean,
```
Comment (fruch) [inline sdcm/sct_config.py:1476]: I would suggest adding some verification step, to make sure we don't enable it for backend that are not tested yet.
or that we won't support for now, like k8s one ?

**Comment by soyacz** [comment]
longer duration failed due some manager nemesis issue.
It managed to execute 44 nemesis - none shown problems with dns name.
argus: https://argus.scylladb.com/test/c38d2050-1e12-479a-8942-40cfaae44859/runs?additionalRuns[]=7aac12d4-ef23-4927-9411-8a5eb232ab88

Only thing I found is to make also manager to communicate with dns name instead of ip.

File: sdcm/cluster.py (hunk 1 above)
Comment (soyacz) [reply sdcm/cluster.py:3952]: yes, and I found this naming is really messed up all around.
I'll try to make it right.

**Comment by soyacz** [comment]
@fgelcer which test do you think we could add  this`use_dns_names` setting to? Some tier1 one?

File: test-cases/longevity/longevity-mini-test-1h.yaml [hunk 4]
```diff
@@ -1,5 +1,5 @@
-test_duration: 180
-stress_cmd: ["cassandra-stress write cl=QUORUM duration=90m -schema 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native -rate threads=50 -pop seq=1..10000000 -log interval=5"]
+test_duration: 1000
+stress_cmd: ["cassandra-stress write cl=QUORUM duration=900m -schema 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native -rate threads=2 -pop seq=1..10000000 -log interval=5"]
```
Comment (fgelcer) [inline test-cases/longevity/longevity-mini-test-1h.yaml:2]: i guess you changed this just to debug purpose, so, once you are done, please, revert the changes in here

File: sdcm/cluster.py [hunk 5]
```diff
@@ -250,6 +250,7 @@ def __init__(self, name, parent_cluster, ssh_login_info=None, base_logdir=None,
 
         self.remoter: Optional[RemoteCmdRunnerBase] = None
 
+        self._use_dns_names: bool = parent_cluster.params.get('use_dns_names') or False
```
Comment (fgelcer) [inline sdcm/cluster.py:253]: we usually set it in one of the config files, either specific to backend, like `aws_config.yaml` or `test_default.yaml`, so setting it in here like this, will be much harder to understand the defaults, when someone will decide to change it

**Review by fgelcer — COMMENTED** [review]
2 little comments

**Comment by fgelcer** [comment]
> @fgelcer which test do you think we could add this`use_dns_names` setting to? Some tier1 one?

why not to run in all of them? i believe that, most of access in the field are made using DNS, instead of IP addresses... so we can keep few tests with private, and few tests with public IPs, but test mainly DNS... @fruch @roydahan , WDYT?

**Comment by roydahan** [comment]
> > @fgelcer which test do you think we could add this`use_dns_names` setting to? Some tier1 one?
> 
> why not to run in all of them? i believe that, most of access in the field are made using DNS, instead of IP addresses... so we can keep few tests with private, and few tests with public IPs, but test mainly DNS... @fruch @roydahan , WDYT?

Because it's not the case.

**Comment by fruch** [comment]
> > > @fgelcer which test do you think we could add this`use_dns_names` setting to? Some tier1 one?
> > 
> > why not to run in all of them? i believe that, most of access in the field are made using DNS, instead of IP addresses... so we can keep few tests with private, and few tests with public IPs, but test mainly DNS... @fruch @roydahan , WDYT?
> 
> Because it's not the case.

So randomly select one of the tier1 cases ? :)

Or we'll duplicate some test case and run it in releases only ?

**Comment by fruch** [comment]
> > > > @fgelcer which test do you think we could add this`use_dns_names` setting to? Some tier1 one?
> > > 
> > > 
> > > why not to run in all of them? i believe that, most of access in the field are made using DNS, instead of IP addresses... so we can keep few tests with private, and few tests with public IPs, but test mainly DNS... @fruch @roydahan , WDYT?
> > 
> > 
> > Because it's not the case.
> 
> So randomly select one of the tier1 cases ? :)
> 
> Or we'll duplicate some test case and run it in releases only ?

@roydahan @fgelcer, we need an answer

**Comment by roydahan** [comment]
Pick one of the tier1 cases.

**Comment by soyacz** [comment]
I added it to 3day longevity

**Comment by soyacz** [comment]
@fruch pylint passes - for it for simplicity I added default implementation of dns name - which will not work for some backends (for some will) - anyway we run it only in AWS and in case we need support for GCE/Azure we can implement proper methods.

unit tests fixed

File: sdcm/cluster_aws.py [hunk 6]
```diff
@@ -513,6 +513,18 @@ def external_address(self):
         else:
             return self._instance.private_ip_address
 
+    @cached_property
```
Comment (fruch) [inline sdcm/cluster_aws.py:516]: I wonder if we need to uncache it on reboot of a node

**Review by fruch — APPROVED** [review]
LGTM.

One comment about caching

File: sdcm/cluster_aws.py (hunk 6 above)
Comment (soyacz) [reply sdcm/cluster_aws.py:516]: in that case we should raise some issue - as it would break cluster network.

Comment (fruch) [reply sdcm/cluster_aws.py:516]: seems like we are using by default the IP name 
```
Hostname type
IP name: ip-10-4-0-64.eu-west-1.compute.internal
```
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-naming.html

there is one case where we stop and start instance, it's in `AWSNode.restart()` it's part of nemesis that is long disabled.
but we are calling there `refresh_ip_address`, that clears all ip related caches, since we've seen cases the public ip address is changing, I hope that local ip address never changes, cause our hostname are matching it.

I still think we should clear the cache in `refresh_ip_address` calls, and maybe generate error event if it was changing.

Comment (soyacz) [reply sdcm/cluster_aws.py:516]: regardless, dns name should never change

Comment (soyacz) [reply sdcm/cluster_aws.py:516]: in that case, it may happen. Added proper cache invalidation in `refresh_ip_address` (btw. fixing `cql_ip_address` invalidation along)

**Review by fruch — APPROVED** [review]
LGTM

**Comment by fruch** [comment]
@soyacz do you have a short run of tier1 test it was added to ? 
just a quick proof it's not completely breaking it ?

and then I think we can merge this one

**Comment by soyacz** [comment]
> @soyacz do you have a short run of tier1 test it was added to ? just a quick proof it's not completely breaking it ?
> 
> and then I think we can merge this one

https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/longevity-50gb-3days-test/2/ - runs ok (got into nemesis part and run few of them). I aborted it after verification.

**Comment by soyacz** [comment]
ready for merge I think

**Review by fruch — APPROVED** [review]
LGTM
